services/arrow-flight/src/main/java/com/dremio/service/flight/DremioFlightSqlProducer.java

```diff
@@ -78,15 +107,29 @@ public void createPreparedStatement(
     ActionCreatePreparedStatementRequest actionCreatePreparedStatementRequest,
     CallContext callContext,
     StreamListener<Result> streamListener) {
-    throw CallStatus.UNIMPLEMENTED.withDescription("createPreparedStatement not supported.").toRuntimeException();
+
+    final FlightDescriptor flightDescriptor =
+      FlightDescriptor.command(actionCreatePreparedStatementRequest.getQuery().getBytes(StandardCharsets.UTF_8));
+
+    final CallHeaders headers = retrieveHeadersFromCallContext(callContext);
+    final UserSession session = sessionsManager.getUserSession(callContext.peerIdentity(), headers);
+    final FlightPreparedStatement flightPreparedStatement = flightWorkManager
+      .createPreparedStatement(flightDescriptor, callContext::isCancelled, session);
+
+    flightPreparedStatementCache.put(flightPreparedStatement.getServerHandle(), flightPreparedStatement);
+
+    final ActionCreatePreparedStatementResult action = flightPreparedStatement.createAction();
+
+    streamListener.onNext(new Result(pack(action).toByteArray()));
+    streamListener.onCompleted();
   }
 
   @Override
   public void closePreparedStatement(
     ActionClosePreparedStatementRequest actionClosePreparedStatementRequest,
     CallContext callContext,
     StreamListener<Result> listener) {
-    throw CallStatus.UNIMPLEMENTED.withDescription("closePreparedStatement not supported.").toRuntimeException();
+    // No need to close PreparedStatement
```

Comment:
? Shouldn't you remove it from the cache?

Reply:
You are right! Fixed